### PR TITLE
INS-34: global resolver

### PR DESCRIPTION
**Pull request by megge-dream**
closes INS-34

**- What I did**
Resolver interface now gets and returns interface{}
Add InitiateResolver to smart contract & domain
Add tests for InitiateResolver
Fix register method for reference domain
Fixes of tests for reference domain

**- How to verify it**
Run tests

**- Description for the changelog**
Add GlobalResolver instance

### CODE REVIEW

File: genesis/model/contract/smart_contract.go [hunk 1]
```diff
@@ -37,16 +39,31 @@ type BaseSmartContract struct {
...
 		CompositeMap: make(map[string]object.Composite),
 		ChildStorage: storage.NewMapStorage(),
 		Parent:       parent,
 	}
+	sc.InitiateResolver(sc)
+	return &sc
+}
+
+// InitiateResolver create resolver.
+func (sc *BaseSmartContract) InitiateResolver(self interface{}) {
```
Comment (LinnikD) [inline genesis/model/contract/smart_contract.go]: InitResolver name will be better

File: genesis/model/contract/smart_contract.go [hunk 2]
```diff
@@ -37,16 +39,31 @@ type BaseSmartContract struct {
...
+	if sc.Resolver != nil {
+		return
+	}
+	scAsParent, ok := self.(object.Parent)
+	if !ok {
```
Comment (LinnikD) [inline genesis/model/contract/smart_contract.go]: NewHandler can handle this case. No need to pass self to method call as a parameter interface{}

File: genesis/model/contract/smart_contract.go (hunk 1 above)
Comment (LinnikD) [inline genesis/model/contract/smart_contract.go]: You can also rename it to GetResolver() and make resolver field private. GetResolver method will act like singleton - return instance or create it if necessary

File: genesis/model/resolver/child.go [hunk 3]
```diff
@@ -47,7 +50,11 @@ func (r *childResolver) GetObject(ref *object.Reference, classID string) (object
 		return nil, fmt.Errorf("object is not Proxy")
 	}
 
-	if proxy.GetClassID() != classID {
+	classIDString, ok := classID.(string)
```
Comment (LinnikD) [inline genesis/model/resolver/child.go]: it is bad style to use types in names. Rename to classID. previous classID can be named like cls

File: genesis/model/resolver/context.go [hunk 4]
```diff
@@ -51,19 +54,24 @@ func (r *contextResolver) GetObject(ref *object.Reference, classID string) (obje
 	}
 
 	for proxy.GetClassID() == class.ReferenceID {
-		contextHolderWithChildInterface, ok := contextHolder.(object.Child)
-		if !ok {
+		contextHolderWithChildInterface, isOk := contextHolder.(object.Child)
```
Comment (LinnikD) [inline genesis/model/resolver/context.go]: `ok` was better name

File: genesis/model/resolver/context.go [hunk 5]
```diff
@@ -51,19 +54,24 @@ func (r *contextResolver) GetObject(ref *object.Reference, classID string) (obje
...
-		proxy, err = contextResolver.GetObject(proxy.(*object.Reference), classID)
+		newProxy, err := contextResolver.GetObject(proxy, classID)
 		if err != nil {
 			return nil, err
 		}
+		proxy = newProxy.(object.Proxy)
 	}
 
-	if proxy.GetClassID() != classID {
+	classIDString, ok := classID.(string)
```
Comment (LinnikD) [inline genesis/model/resolver/context.go]: simple names is better

File: genesis/model/resolver/global.go [hunk 6]
```diff
@@ -24,31 +24,50 @@ import (
...
-	if proxy.GetClassID() != classID {
+	parent, ok := parentProxy.(object.Parent)
+	if !ok {
+		return nil, fmt.Errorf("object with domain `%s` can not have children", ref.Domain)
+	}
+	proxy, err := parent.GetChild(ref.Record)
+	if err != nil {
+		return nil, err
+	}
+	classIDString, ok := classID.(string)
```
Comment (LinnikD) [inline genesis/model/resolver/global.go]: here `classIDString ` also need to be renamed

File: genesis/model/resolver/handler.go [hunk 7]
```diff
@@ -22,23 +22,28 @@ import (
...
+	return &Handler{
 		globalResolver:  GlobalResolver,
 		childResolver:   newChildResolver(p),
 		contextResolver: newContextResolver(p),
 	}
 }
 
 // GetObject resolve object by its reference and return its proxy.
-func (r *resolverHandler) GetObject(ref *object.Reference, classID string) (object.Proxy, error) {
+func (r *Handler) GetObject(reference interface{}, classID interface{}) (interface{}, error) {
```
Comment (LinnikD) [inline genesis/model/resolver/handler.go]: Add description to doc: why we use interface{} types for reference, class and proxy

File: genesis/model/resolver/global.go [hunk 8]
```diff
@@ -24,31 +24,50 @@ import (
...
+	if proxy.GetClassID() != classIDString {
 		return nil, fmt.Errorf("instance class is not `%s`", classID)
 	}
 	return proxy, nil
 }
 
+// SetMap set globalInstanceMap into globalResolver
+func (r *globalResolver) SetMap(globalInstanceMap *map[string]object.Proxy) error {
+	if r.globalInstanceMap != nil {
+		return fmt.Errorf("map was already set")
```
Comment (LinnikD) [inline genesis/model/resolver/global.go]: Is it really error case? If map already set we can act normally. Will it be error case in future implementation?

File: genesis/public/core/reference.go [hunk 9]
```diff
@@ -33,13 +33,16 @@ type ReferenceDomain interface {
 	// Base domain implementation.
 	domain.Domain
 	// RegisterReference is used to publish new global references
-	RegisterReference(*object.Reference) (string, error)
+	RegisterReference(*object.Reference, string) (string, error)
 	// ResolveReference provides reference instance from record
 	ResolveReference(string) (*object.Reference, error)
+	// SetMap set globalResolverMap for references register/resolving
+	SetMap(globalInstanceMap *map[string]object.Proxy) error
```
Comment (LinnikD) [inline genesis/public/core/reference.go]: `SetMap` is not good context name here. Maybe `InitGlobalResolver` is better?

File: genesis/model/resolver/context.go (hunk 4 above)
Comment (megge-dream) [reply genesis/model/resolver/context.go]: declaration of ```ok``` shadows declaration at line 40 (in this function but outside of cycle)

Comment (LinnikD) [reply genesis/model/resolver/context.go]: `isChild` then ;)

File: genesis/public/core/reference.go (hunk 9 above)
Comment (megge-dream) [reply genesis/public/core/reference.go]: Its not about init global resolver, its about set inner reference domain map at the same memory point as global resolver map. InitGlobalMap mb?

File: genesis/public/core/reference.go [hunk 10]
```diff
@@ -96,8 +117,8 @@ func newReferenceDomainProxy(parent object.Parent) *referenceDomainProxy {
 }
 
 // RegisterReference proxy call for instance method.
-func (rdp *referenceDomainProxy) RegisterReference(address *object.Reference) (string, error) {
-	return rdp.instance.RegisterReference(address)
+func (rdp *referenceDomainProxy) RegisterReference(address *object.Reference, classID string) (string, error) {
+	return rdp.instance.RegisterReference(address, classID)
 }
 
 // ResolveReference proxy call for instance method.
```
Comment (dmitry-panchenko) [inline genesis/public/core/reference.go]: ResolveReference is a ...

File: genesis/public/core/reference.go [hunk 11]
```diff
@@ -120,6 +141,11 @@ func (rdp *referenceDomainProxy) GetClassID() string {
 	return class.ReferenceDomainID
 }
 
+// InitGlobalMap proxy call for instance method.
```
Comment (dmitry-panchenko) [inline genesis/public/core/reference.go]: InitGlobalMap is a

**Pull request merged**